### DISCUSSION TITLE
GPytorch how to do evaluation 2 times with 1 GP model in the same custom likelihood?

### DISCUSSION THREAD

**Discussion by TunaYuce** [cornellius-gp/gpytorch#2448] [Q&A]
I have data that includes features of 2 pairs and their relationship as a label. These labels show if both pairs are equal to each other less than a threshold in any case or whichever is bigger than the threshold. If pair1 is bigger, label is 0. If they are equal the label is 1. If pair2 is bigger, the label is 2. So my data looks like below;

```
Pair1_x1, Pair1_x2, Pair2_x1, Pair2_x2, Label
10, 20, 30, 40, 2
10, 20, 20, 20, 1
20, 30, 60, 10, 0
...
```

I want to create a custom likelihood for a gp solution. The thing is, I should be able to somehow get predictions for pair1 and pair2 separately from the same model and calculate the probabilities together with an appropriate loss. My problem is, I could not find any example or a solution for doing that. I created my model with ApproximateGP in the way below:

```
class MyGPModel(ApproximateGP):
    def __init__(self, likelihood, input_dim):
        variational_distribution = CholeskyVariationalDistribution(10)
        inducing_points = torch.zeros(10, input_dim)
        variational_strategy = VariationalStrategy(self, inducing_points, variational_distribution, learn_inducing_locations=True)
        
        super().__init__(variational_strategy)
        self.mean_module = gpytorch.means.ConstantMean()
        self.covar_module = gpytorch.kernels.ScaleKernel(gpytorch.kernels.RBFKernel(ard_num_dims=input_dim))
        self.likelihood = likelihood

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)
```

I can create the custom likelihood in a way like this:

```
class MyLikelihood(gpytorch.likelihoods.Likelihood):
    def __init__(self):
        super().__init__()
        threshold...

    def forward(self, function_samples, **kwargs):
        
        ....
        prob calculations
        find probabilities for labels 0, 1, 2 etc.
        ...
        return probs
```

Here my problem is, how can I send the predictions from both pair1 and pair2 to create a custom likelihood and I am not sure about how to implement the loss part since I can not directly use a built-in function like VariationalELBO since they always get 1 kind of data, only 1 pair in my case. I can maybe call the model for 2 pairs separately like that:

```
output1 = model(X_train_pair1)
output2 = model(X_train_pair2)
```

and then pass the combination of means and covs to the likelihood as a MultivariateNormal but I am not sure how to do it right now. Any ideas about my problem?

Note: I already tried pairwise GP from botorch and but I think it was not exactly suited to my data.
